### PR TITLE
refactor(awful: hotkeys_popup): expose configuration options for a widget instance and use more object-oriented structure for the widget

### LINKED ISSUE #1352: Why does awful.hotkeys_popup use notification_opacity
This seems at least surprising. I wouldn't consider this a notification. Can it use its own beautiful key instead?
https://github.com/awesomeWM/awesome/blob/3fea2db8c5f8cd5a2f3ed463fe57dcdc673c3afc/lib/awful/hotkeys_popup/widget.lua#L338

### LINKED ISSUE #1497: [awful.hotkey_widget] get theme parameters when creating widget's wibox, not when initializing widget object itself
should fix the problem found by #1471 and previous psychons's attempts to figure out the mess with beautiful import order/cyclic deps

### PR DESCRIPTION
**Pull request by actionless**
closes #1352
closes #1497

### CODE REVIEW

**Comment by Elv13** [comment]
Didn't really do a review yet, but the first this I notice is that the `@beautiful` documentation blocks seem to be missing. They are required for the page that list all possible theme variables (and the hotkey own doc page, of course).

**Comment by actionless** [comment]
@Elv13 submitted

File: lib/awful/hotkeys_popup/widget.lua [hunk 1]
```diff
@@ -44,22 +55,71 @@ widget.hide_without_description = true
 widget.merge_duplicates = true
 
 
+--- Widget background color.
+-- @beautiful beautiful.hotkeys_bg
```
Comment (Elv13) [inline lib/awful/hotkeys_popup/widget.lua]: I really feel like I am nitpicking right now, but you should use the "hotkeys" or "hotkeys widget" in the description and add the type to all of them. See https://awesomewm.org/apidoc/documentation/06-appearance.md.html 

Not really important, but coherency, boring predictability and conventions goes a long way when you document a library.

File: lib/awful/hotkeys_popup/widget.lua [hunk 2]
```diff
@@ -100,55 +160,77 @@ function widget.new()
...
+        self.border_width = args.border_width or
+            beautiful.hotkeys_border_width or beautiful.border_width
+        self.border_color = args.border_color or
+            beautiful.hotkeys_border_color or self.fg
+        self.shape = args.shape or beautiful.hotkeys_shape
+        self.modifiers_fg = args.modifiers_fg or
+            beautiful.hotkeys_modifiers_fg or beautiful.bg_minimize or "#555555"
+        self.label_fg = args.label_fg or
+            beautiful.hotkeys_label_fg or self.bg
+        self.opacity = beautiful.notification_opacity or 1
```
Comment (Elv13) [inline lib/awful/hotkeys_popup/widget.lua]: As a couple of people already told us on IRC, this still make little sense. Please add a proper variable for it. You can fallback to `beautiful.notification_opacity` is you thing it's worth it. I would not mind the behavior break if this was changed.

File: lib/awful/hotkeys_popup/widget.lua [hunk 3]
```diff
@@ -100,55 +160,77 @@ function widget.new()
...
+            beautiful.hotkeys_bg or beautiful.bg_normal
+        self.fg = args.fg or
+            beautiful.hotkeys_fg or beautiful.fg_normal
+        self.border_width = args.border_width or
+            beautiful.hotkeys_border_width or beautiful.border_width
+        self.border_color = args.border_color or
+            beautiful.hotkeys_border_color or self.fg
+        self.shape = args.shape or beautiful.hotkeys_shape
+        self.modifiers_fg = args.modifiers_fg or
+            beautiful.hotkeys_modifiers_fg or beautiful.bg_minimize or "#555555"
```
Comment (Elv13) [inline lib/awful/hotkeys_popup/widget.lua]: beautiful.bg_minimize

This also looks like a random choice

File: lib/awful/hotkeys_popup/widget.lua [hunk 4]
```diff
@@ -357,21 +437,21 @@ function widget.new()
...
-            self.current_page = self.current_page - 1
-            self.wibox:set_widget(pages[self.current_page])
+        function widget_obj.page_prev(_self)
+            if _self.current_page == 1 then return end
+            _self.current_page = _self.current_page - 1
+            _self.wibox:set_widget(pages[_self.current_page])
         end
-        function widget_obj:show()
-            self.wibox.visible = true
+        function widget_obj.show(_self)
```
Comment (Elv13) [inline lib/awful/hotkeys_popup/widget.lua]: Why the change? If it was about shadowing `self`, then you should consider moving these methods out of the constructor.

File: themes/default/theme.lua [hunk 5]
```diff
@@ -39,6 +39,7 @@ theme.border_marked = "#91231c"
 -- tooltip_[font|opacity|fg_color|bg_color|border_width|border_color]
 -- mouse_finder_[color|timeout|animate_timeout|radius|factor]
 -- prompt_[fg|bg|fg_cursor|bg_cursor|font]
+-- hotkeys_[bg|fg|border_width|border_color|shape|opacity|modifiers_fg|label_fg|group_margin|font|description_font]
```
Comment (Elv13) [inline themes/default/theme.lua]: Just to note that `opacity` doesn't seem to exist in this revision

**Comment by Elv13** [comment]
Overall it's good. Only some minor comments here and there.

File: lib/awful/hotkeys_popup/widget.lua (hunk 2 above)
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: actually i've documented it in the default theme.lua but forgot to add actually :)

File: lib/awful/hotkeys_popup/widget.lua (hunk 3 above)
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: that value is usually looking the best with most of the themes i've tried

in beautiful we don't have anything like `theme.bg_disabled` and that `bg_minimize` is the most closer to represent some inactive/disabled info (or like in this case to make the labels with that color less accentiated)

File: lib/awful/hotkeys_popup/widget.lua (hunk 4 above)
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: yup, because of that, i'll move them then

File: lib/awful/hotkeys_popup/widget.lua (hunk 1 above)
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: for some reason those beautiful docstrings are displayed in apidoc for the hotkeys_popup.widget but not present in 06-appearance

could it be because they're missing `@param`?

File: lib/awful/hotkeys_popup/widget.lua (hunk 4 above)
Comment (actionless) [inline lib/awful/hotkeys_popup/widget.lua]: mm, i think in the place it's more clear

**Comment by actionless** [comment]
is it fine now?

**Comment by actionless** [comment]
@blueyed, @psychon any objections?

**Comment by psychon** [comment]
I did not look too closely, but seems sensible. I trust Elv13 here.
